### ISSUE TITLE
cpp-netlib 0.12.0 build error with visual studio 2015 and boost 1.60

### ISSUE THREAD

**Issue by matsuz** [cpp-netlib/cpp-netlib#648]
I'm sorry for my English.

**my environment:**
+ windows 10 th2 10586.218 x64
+ visual studio community 2015 update2
+ boost 1.60
+ cpp-netlib-0.12.0-final

---
I build boost with this command in x86 native mode:
```cmd
b2 install --prefix=C:\Users\Samuel\Desktop\builds\target --toolset=msvc --without-python link=static runtime-link=shared debug release
```
---
and I generate `cpp-netlib.sln` with this cmake command:
```cmd
mkdir cpn-build
cd cpn-build
cmake ..\cpp-netlib-0.12.0-final -G "Visual Studio 14" -DBOOST_ROOT=C:\Users\Samuel\Desktop\builds\target
```
---
and then, I tried to build cpp-netlib in visual studio.
But, in build project `cppnetlib-server-parsers`, vs2015 tips:
> C:\Users\Samuel\Desktop\builds\target\include\boost-1_60\boost/type_traits/is_const.hpp(37): error C2872: “true_type”: ambiguous symbol

and build failed eventually.

Am I build it in wrong way? How can I build it correctly?

**Comment by snap9** [comment]
I'm having exactly this problem.

Anyone found a solution yet?

**Comment by ytong3** [comment]
If you can live without 0.12, you may want to try 0.11.   
  
  
 Sent using   Email (https://itunes.apple.com/app/apple-store/id922793622?pt=814382&ct=Email&mt=8)
  
  
  
  

  
  
>

**Comment by deanberris** [comment]
I don't have a solution yet because I don't have access to MSVC 2015 -- and I don't know where exactly to start here. Any tips for debugging this would be really helpful.

**Comment by snap9** [comment]
@deanberris I downloaded VS2015 for free on Microsoft's page. It's right here: https://www.visualstudio.com/en-us/products/visual-studio-express-vs.aspx (I clicked on Download under "Express for Desktop").

**Comment by burannah** [comment]
As a fast & dirty workaround you can change that `true_type`@ is_const.hpp:37 to `std::true_type` to avoid ambiguous declaration.
It's (almost) the same type.

**Comment by snap9** [comment]
@burannah That quick and dirty workaround seems to work. Thanks! @deanberris Maybe you guys can add the change in the official cpp-netlib? (See my above comment about getting VS2015 for free.)

**Comment by enricodetoma** [comment]
@burannah but is_const.hpp:37 is in Boost, not in cpp-netlib!

**Comment by enricodetoma** [comment]
I have create pull request #685 https://github.com/cpp-netlib/cpp-netlib/pull/685 to solve the compilation problem with Visual Studio 2015

**Comment by enricodetoma** [comment]
I can confirm that this issue can be closed for milestone 0.13 too.

**Comment by mjfryc** [comment]
It was needed to compile `cppnetlib-server-parsers`.

Solution: in my case I had to modify:
`[cpp-netlib-0.11.2_dir]\boost\network\protocol\http\server\impl\parsers.ipp`
by:
- Extract function `BOOST_NETWORK_INLINE void parse_version` to anonymous namespace and rename it e.g. to `parse_version_hack`.
- In "Original" `parse_version` function do NOT use `using namespace boost::spirit::qi;` and delegate this function to anonymous namespace function.

Below example (sorry for identation lack):
`namespace
{
    BOOST_NETWORK_INLINE void parse_headers_hack(
        std::string const& input, std::vector<boost::network::http::request_header_narrow>& container) {
      using namespace boost::spirit::qi;
      boost::u8_to_u32_iterator<std::string::const_iterator> begin = input.begin(), end = input.end();
      typedef boost::spirit::qi::as<boost::spirit::traits::u32_string> as_u32_string;
      parse(begin, end,
            *(+((alnum | punct) - ':') >> lit(": ") >>
              as_u32_string()[+((unicode::alnum | space | punct) - '\r' - '\n')] >>
              lit("\r\n")) >>
                lit("\r\n"),
            container);
    }
}

namespace boost {
namespace network {
namespace http {

BOOST_NETWORK_INLINE void parse_version(
    std::string const& partial_parsed, boost::fusion::tuple<boost::uint8_t, boost::uint8_t>& version_pair)
    {
        boost::spirit::qi::parse
        (
            partial_parsed.begin(), 
            partial_parsed.end(), 
            (
                boost::spirit::qi::lit("HTTP/") >> boost::spirit::qi::ushort_ >> '.' >> boost::spirit::qi::ushort_
            ),
            version_pair
        );
    }`
